drivers/sensor/st/lsm6dsv16x/lsm6dsv16x.c

```diff
@@ -282,6 +283,229 @@ static int lsm6dsv16x_accel_wake_duration_set(const struct device *dev,
 }
 
+#ifdef CONFIG_LSM6DSV16X_SELF_TEST
+/* Self test limits. */
+#define MIN_ST_LIMIT_mg		50.0f
+#define MAX_ST_LIMIT_mg		1700.0f
+#define MIN_ST_LIMIT_mdps	150000.0f
+#define MAX_ST_LIMIT_mdps	700000.0f
+
+/*
+ * Accelerometer Self Test
+ */
+static int lsm6dsv16x_accel_self_test(const struct device *dev)
+{
+	const struct lsm6dsv16x_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lsm6dsv16x_data_ready_t drdy;
+	int16_t data_raw[3];
+	float_t val_st_off[3];
+	float_t val_st_on[3];
+	float_t test_val[3];
+	uint8_t i, j, st_result;
+
+	/* Set Output Data Rate */
+	lsm6dsv16x_xl_data_rate_set(ctx, LSM6DSV16X_ODR_AT_60Hz);
+
+	/* Set full scale */
+	lsm6dsv16x_xl_full_scale_set(ctx, LSM6DSV16X_4g);
+
+	/* Wait stable output */
+	k_msleep(100);
+
+	/* Check if new value available */
+	do {
+		lsm6dsv16x_flag_data_ready_get(ctx, &drdy);
+	} while (!drdy.drdy_xl);
+
+	/* Read dummy data and discard it */
+	lsm6dsv16x_acceleration_raw_get(ctx, data_raw);
+
+	/* Read 5 sample and get the average vale for each axis */
+	memset(val_st_off, 0x0, 3 * sizeof(float));
+
+	for (i = 0; i < 5; i++) {
+		/* Check if new value available */
+		do {
+			lsm6dsv16x_flag_data_ready_get(ctx, &drdy);
+		} while (!drdy.drdy_xl);
+
+		/* Read data and accumulate the mg value */
+		lsm6dsv16x_acceleration_raw_get(ctx, data_raw);
+
+		for (j = 0; j < 3; j++) {
+			val_st_off[j] += lsm6dsv16x_from_fs4_to_mg(data_raw[j]);
+		}
+	}
+
+	/* Calculate the mg average values */
+	for (i = 0; i < 3; i++) {
+		val_st_off[i] /= 5.0f;
+	}
+
+	/* Enable Self Test negative */
+	lsm6dsv16x_xl_self_test_set(ctx, LSM6DSV16X_XL_ST_NEGATIVE);
+
+	/* Wait stable output */
+	k_msleep(100);
+
+	/* Check if new value available */
+	do {
+		lsm6dsv16x_flag_data_ready_get(ctx, &drdy);
+	} while (!drdy.drdy_xl);
+
+	/* Read dummy data and discard it */
+	lsm6dsv16x_acceleration_raw_get(ctx, data_raw);
+	/* Read 5 sample and get the average vale for each axis */
+	memset(val_st_on, 0x00, 3 * sizeof(float));
+
+	for (i = 0; i < 5; i++) {
+		/* Check if new value available */
+		do {
+			lsm6dsv16x_flag_data_ready_get(ctx, &drdy);
+		} while (!drdy.drdy_xl);
+
+		/* Read data and accumulate the mg value */
+		lsm6dsv16x_acceleration_raw_get(ctx, data_raw);
+
+		for (j = 0; j < 3; j++) {
+			val_st_on[j] += lsm6dsv16x_from_fs4_to_mg(data_raw[j]);
+		}
+	}
+
+	/* Calculate the mg average values */
+	for (i = 0; i < 3; i++) {
+		val_st_on[i] /= 5.0f;
+	}
+
+	/* Calculate the mg values for self test */
+	for (i = 0; i < 3; i++) {
+		test_val[i] = fabsf((val_st_on[i] - val_st_off[i]));
+	}
+
+	/* Check self test limit */
+	st_result = 0;
+	for (i = 0; i < 3; i++) {
+		if ((test_val[i] < MIN_ST_LIMIT_mg) ||
+		    (test_val[i] > MAX_ST_LIMIT_mg)) {
+			st_result = -1;
+			break;
+		}
+	}
+
+	/* Disable Self Test */
+	lsm6dsv16x_xl_self_test_set(ctx, LSM6DSV16X_XL_ST_DISABLE);
+
+	/* Disable sensor. */
+	lsm6dsv16x_xl_data_rate_set(ctx, LSM6DSV16X_ODR_OFF);
+
```

Comment:
As a suggestion, I'd add a `LOG_INF` print here and in `lsm6dsv16x_gyro_self_test` to report the self-test status. We already print stuff like `chip ID` when a device is being probed and it's really convenient to do a quick check in e.g. RTT logs